File: c/qc_encoder_all/generate_constants.py
from __future__ import annotations

from pathlib import Path


CONFIGS = [
    {
        "selection": 1,
        "name": "rate_1_2_1k",
        "g_path": "../build_g/G_1_2_1024.mat",
        "info_length": 1024,
        "transmitted_length": 2048,
        "full_length": 2560,
        "row_blocks": 8,
        "col_blocks": 8,
        "block_size": 128,
    },
    {
        "selection": 2,
        "name": "rate_1_2_4k",
        "g_path": "../build_g/G_1_2_4096.mat",
        "info_length": 4096,
        "transmitted_length": 8192,
        "full_length": 10240,
        "row_blocks": 8,
        "col_blocks": 8,
        "block_size": 512,
    },
    {
        "selection": 3,
        "name": "rate_1_2_16k",
        "g_path": "../build_g/G_1_2_16384.mat",
        "info_length": 16384,
        "transmitted_length": 32768,
        "full_length": 40960,
        "row_blocks": 8,
        "col_blocks": 8,
        "block_size": 2048,
    },
    {
        "selection": 4,
        "name": "rate_2_3_1k",
        "g_path": "../build_g/G_2_3_1024.mat",
        "info_length": 1024,
        "transmitted_length": 1536,
        "full_length": 1792,
        "row_blocks": 16,
        "col_blocks": 8,
        "block_size": 64,
    },
    {
        "selection": 5,
        "name": "rate_2_3_4k",
        "g_path": "../build_g/G_2_3_4096.mat",
        "info_length": 4096,
        "transmitted_length": 6144,
        "full_length": 7168,
        "row_blocks": 16,
        "col_blocks": 8,
        "block_size": 256,
    },
    {
        "selection": 6,
        "name": "rate_2_3_16k",
        "g_path": "../build_g/G_2_3_16384.mat",
        "info_length": 16384,
        "transmitted_length": 24576,
        "full_length": 28672,
        "row_blocks": 16,
        "col_blocks": 8,
        "block_size": 1024,
    },
    {
        "selection": 7,
        "name": "rate_4_5_1k",
        "g_path": "../build_g/G_4_5_1024.mat",
        "info_length": 1024,
        "transmitted_length": 1280,
        "full_length": 1408,
        "row_blocks": 32,
        "col_blocks": 8,
        "block_size": 32,
    },
    {
        "selection": 8,
        "name": "rate_4_5_4k",
        "g_path": "../build_g/G_4_5_4096.mat",
        "info_length": 4096,
        "transmitted_length": 5120,
        "full_length": 5632,
        "row_blocks": 32,
        "col_blocks": 8,
        "block_size": 128,
    },
    {
        "selection": 9,
        "name": "rate_4_5_16k",
        "g_path": "../build_g/G_4_5_16384.mat",
        "info_length": 16384,
        "transmitted_length": 20480,
        "full_length": 22528,
        "row_blocks": 32,
        "col_blocks": 8,
        "block_size": 512,
    },
]

SUPPORTED_SELECTIONS = {1, 2, 4, 5, 7, 8}


def load_octave_sparse_text_mat(
    path: Path,
    expected_rows: int,
    expected_cols: int,
    info_length: int,
    transmitted_length: int,
    col_blocks: int,
    block_size: int,
) -> list[list[int]]:
    rows = -1
    cols = -1
    nnz_expected = -1
    nnz_loaded = 0
    variable_name = None
    target_rows = {row_block * block_size for row_block in range(expected_rows // block_size)}
    block_rows: list[list[int]] | None = None

    with path.open("r", encoding="ascii") as handle:
        for raw_line in handle:
            line = raw_line.strip()
            if not line:
                continue
            if line.startswith("# name:"):
                variable_name = line.split(":", 1)[1].strip()
                continue
            if line.startswith("# type:"):
                if "sparse matrix" not in line:
                    raise ValueError(f"{path} is not a sparse matrix file")
                continue
            if line.startswith("# nnz:"):
                nnz_expected = int(line.split(":", 1)[1].strip())
                continue
            if line.startswith("# rows:"):
                rows = int(line.split(":", 1)[1].strip())
                continue
            if line.startswith("# columns:"):
                cols = int(line.split(":", 1)[1].strip())
                block_rows = [[0] * col_blocks for _ in range(rows)]
                continue
            if line.startswith("#"):
                continue
            if block_rows is None:
                raise ValueError("matrix entries were seen before dimensions were parsed")

            row_text, col_text, value_text = line.split()
            row_index = int(row_text) - 1
            col_index = int(col_text) - 1
            value = int(value_text)
            if value:
                nnz_loaded += 1
                if row_index in target_rows and info_length <= col_index < transmitted_length:
                    transmitted_col = col_index - info_length
                    col_block = transmitted_col // block_size
                    bit_index = transmitted_col % block_size
                    block_rows[row_index][col_block] |= 1 << bit_index

    if variable_name != "G":
        raise ValueError(f"expected variable G in {path}, got {variable_name!r}")
    if rows != expected_rows or cols != expected_cols:
        raise ValueError(f"unexpected G dimensions {rows} x {cols}, expected {expected_rows} x {expected_cols}")
    if nnz_expected >= 0 and nnz_expected != nnz_loaded:
        raise ValueError(f"nnz mismatch in {path}: expected {nnz_expected}, loaded {nnz_loaded}")
    assert block_rows is not None
    return block_rows


def rotate_first_row_bits(first_row_bits: int, shift: int, block_size: int) -> int:
    mask = (1 << block_size) - 1
    shift %= block_size
    if shift == 0:
        return first_row_bits & mask
    return ((first_row_bits << shift) | (first_row_bits >> (block_size - shift))) & mask


def validate_circulant(
    block_rows: list[list[int]],
    block_size: int,
    row_block: int,
    col_block: int,
) -> int:
    row_start = row_block * block_size
    first_row_bits = block_rows[row_start][col_block]
    return first_row_bits


def bits_to_words(bits_value: int, block_words: int) -> list[int]:
    words = [0] * block_words
    for word_index in range(block_words):
        words[word_index] = (bits_value >> (word_index * 64)) & ((1 << 64) - 1)
    return words


def emit_header(config_rows: list[tuple[dict[str, int | str], list[int]]], output_path: Path) -> None:
    lines: list[str] = []
    lines.append("#ifndef QC_ENCODER_CONSTANTS_H")
    lines.append("#define QC_ENCODER_CONSTANTS_H")
    lines.append("")
    lines.append("#include <stdint.h>")
    lines.append("")

    for config, flat_words in config_rows:
        selection = int(config["selection"])
        lines.append(
            f"static const uint64_t qc_circulant_first_rows_cfg_{selection}[] = {{"
        )
        for index, word in enumerate(flat_words):
            suffix = "," if index + 1 < len(flat_words) else ""
            lines.append(f"    UINT64_C(0x{word:016x}){suffix}")
        lines.append("};")
        lines.append("")

    lines.append("static const qc_encoder_config k_qc_encoder_configs[] = {")
    for config, _flat_words in config_rows:
        selection = int(config["selection"])
        transmitted_length = int(config["transmitted_length"])
        info_length = int(config["info_length"])
        transmitted_parity_length = transmitted_length - info_length
        block_words = (int(config["block_size"]) + 63) // 64
        lines.append("    {")
        lines.append(f"        {selection},")
        lines.append(f"        \"{config['name']}\",")
        lines.append(f"        \"{config['g_path']}\",")
        lines.append(f"        {info_length},")
        lines.append(f"        {transmitted_length},")
        lines.append(f"        {int(config['full_length'])},")
        lines.append(f"        {int(config['row_blocks'])},")
        lines.append(f"        {int(config['col_blocks'])},")
        lines.append(f"        {int(config['block_size'])},")
        lines.append(f"        {block_words},")
        lines.append(f"        {transmitted_parity_length},")
        lines.append(f"        qc_circulant_first_rows_cfg_{selection}")
        lines.append("    },")
    lines.append("};")
    lines.append("")
    lines.append(
        "static const int k_qc_encoder_config_count = (int)(sizeof(k_qc_encoder_configs) / sizeof(k_qc_encoder_configs[0]));"
    )
    lines.append("")
    lines.append("#endif")
    lines.append("")
    output_path.write_text("\n".join(lines), encoding="ascii")


def main() -> int:
    script_dir = Path(__file__).resolve().parent
    config_rows: list[tuple[dict[str, int | str], list[int]]] = []

    for config in CONFIGS:
        if int(config["selection"]) not in SUPPORTED_SELECTIONS:
            continue
        info_length = int(config["info_length"])
        transmitted_length = int(config["transmitted_length"])
        full_length = int(config["full_length"])
        row_blocks = int(config["row_blocks"])
        col_blocks = int(config["col_blocks"])
        block_size = int(config["block_size"])
        block_words = (block_size + 63) // 64

        if row_blocks * block_size != info_length:
            raise ValueError(f"config {config['selection']} has inconsistent row block dimensions")
        if col_blocks * block_size != transmitted_length - info_length:
            raise ValueError(f"config {config['selection']} has inconsistent column block dimensions")

        block_row_values = load_octave_sparse_text_mat(
            script_dir / str(config["g_path"]),
            info_length,
            full_length,
            info_length,
            transmitted_length,
            col_blocks,
            block_size,
        )
        flat_words: list[int] = []
        for row_block in range(row_blocks):
            for col_block in range(col_blocks):
                first_row_bits = validate_circulant(block_row_values, block_size, row_block, col_block)
                flat_words.extend(bits_to_words(first_row_bits, block_words))

        print(
            f"Validated config {config['selection']}: {config['name']} "
            f"({row_blocks}x{col_blocks} blocks, {block_size} bits)"
        )
        config_rows.append((config, flat_words))

    emit_header(config_rows, script_dir / "qc_encoder_constants.h")
    print("Wrote constants header to qc_encoder_constants.h")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: c/qc_encoder_all/generate_constants.c
/*
 * generate_constants.c
 *
 * Exact C equivalent of your Python generator.
 * Fully compatible with qc_encoder.h.
 *
 * Compile:
 *   gcc -O2 -std=c11 generate_constants.c -o generate_constants
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#define MAX_LINE 512
#define MAX_CONFIGS 9

/* ============================================================
 * INPUT CONFIG STRUCTURE (internal only)
 * ============================================================ */

typedef struct
{
    int selection;
    const char *name;
    const char *g_path;
    int info_length;
    int transmitted_length;
    int full_length;
    int row_blocks;
    int col_blocks;
    int block_size;
} qc_config_input;

/* ============================================================
 * SUPPORTED CONFIGS
 * ============================================================ */

static const qc_config_input CONFIGS[] =
{
    {1,"rate_1_2_1k","../build_g/G_1_2_1024.mat",1024,2048,2560,8,8,128},
    {2,"rate_1_2_4k","../build_g/G_1_2_4096.mat",4096,8192,10240,8,8,512},
    {3,"rate_1_2_16k","../build_g/G_1_2_16384.mat",16384,32768,40960,8,8,2048},

    {4,"rate_2_3_1k","../build_g/G_2_3_1024.mat",1024,1536,1792,16,8,64},
    {5,"rate_2_3_4k","../build_g/G_2_3_4096.mat",4096,6144,7168,16,8,256},
    {6,"rate_2_3_16k","../build_g/G_2_3_16384.mat",16384,24576,28672,16,8,1024},

    {7,"rate_4_5_1k","../build_g/G_4_5_1024.mat",1024,1280,1408,32,8,32},
    {8,"rate_4_5_4k","../build_g/G_4_5_4096.mat",4096,5120,5632,32,8,128},
    {9,"rate_4_5_16k","../build_g/G_4_5_16384.mat",16384,20480,22528,32,8,512}
};

static const int CONFIG_COUNT =
    sizeof(CONFIGS)/sizeof(CONFIGS[0]);

/* ============================================================
 * LOAD OCTAVE SPARSE MATRIX
 * ============================================================ */

static uint64_t *load_octave_sparse_text_mat(
    const qc_config_input *cfg,
    int *out_block_words)
{
    FILE *fp = fopen(cfg->g_path,"r");
    if(!fp)
    {
        fprintf(stderr,"Cannot open %s\n",cfg->g_path);
        return NULL;
    }

    int rows=-1, cols=-1;
    int block_words = (cfg->block_size + 63)/64;
    *out_block_words = block_words;

    uint64_t *block_rows = NULL;

    char line[MAX_LINE];

    while(fgets(line,sizeof(line),fp))
    {
        if(line[0]=='#')
        {
            if(strncmp(line,"# rows:",7)==0)
                rows = atoi(line+7);

            else if(strncmp(line,"# columns:",10)==0)
            {
                cols = atoi(line+10);

                block_rows = calloc(
                    rows *
                    cfg->col_blocks *
                    block_words,
                    sizeof(uint64_t));

                if(!block_rows)
                {
                    fclose(fp);
                    return NULL;
                }
            }
            continue;
        }

        if(!block_rows)
            continue;

        int row,col,value;

        if(sscanf(line,"%d %d %d",&row,&col,&value)!=3)
            continue;

        row--; col--;

        if(!value)
            continue;

        if((row % cfg->block_size)==0 &&
           col>=cfg->info_length &&
           col<cfg->transmitted_length)
        {
            int transmitted_col = col - cfg->info_length;
            int col_block = transmitted_col / cfg->block_size;
            int bit_index = transmitted_col % cfg->block_size;

            int word_index = bit_index / 64;
            int bit_offset = bit_index % 64;

            block_rows[
                row * cfg->col_blocks * block_words +
                col_block * block_words +
                word_index
            ] |= ((uint64_t)1 << bit_offset);
        }
    }

    fclose(fp);

    if(rows!=cfg->info_length ||
       cols!=cfg->full_length)
    {
        fprintf(stderr,"Dimension mismatch in %s\n",cfg->g_path);
        free(block_rows);
        return NULL;
    }

    return block_rows;
}

/* ============================================================
 * HEADER GENERATION (MATCHES qc_encoder.h)
 * ============================================================ */

static int emit_header(
    const qc_config_input *configs,
    uint64_t **flat_data,
    int *flat_sizes,
    int *block_words_array,
    int config_count)
{
    FILE *fp = fopen("qc_encoder_constants.h","w");
    if(!fp)
        return -1;

    fprintf(fp,"#ifndef QC_ENCODER_CONSTANTS_H\n");
    fprintf(fp,"#define QC_ENCODER_CONSTANTS_H\n\n");
    fprintf(fp,"#include <stdint.h>\n");
    fprintf(fp,"#include \"qc_encoder.h\"\n\n");

    /* ---- Circulant arrays ---- */

    for(int c=0;c<config_count;c++)
    {
        const qc_config_input *cfg=&configs[c];

        fprintf(fp,
        "static const uint64_t qc_circulant_first_rows_cfg_%d[] = {\n",
        cfg->selection);

        for(int i=0;i<flat_sizes[c];i++)
        {
            fprintf(fp,"    UINT64_C(0x%016llx)%s\n",
                (unsigned long long)flat_data[c][i],
                (i+1<flat_sizes[c])?",":"");
        }

        fprintf(fp,"};\n\n");
    }

    /* ---- Config table ---- */

    fprintf(fp,"static const qc_encoder_config k_qc_encoder_configs[] = {\n");

    for(int c=0;c<config_count;c++)
    {
        const qc_config_input *cfg=&configs[c];
        int block_words = block_words_array[c];
        int parity = cfg->transmitted_length - cfg->info_length;

        fprintf(fp,"    {\n");
        fprintf(fp,"        %d,\n",cfg->selection);
        fprintf(fp,"        \"%s\",\n",cfg->name);
        fprintf(fp,"        \"%s\",\n",cfg->g_path);   // maps to g_matrix_path
        fprintf(fp,"        %d,\n",cfg->info_length);
        fprintf(fp,"        %d,\n",cfg->transmitted_length);
        fprintf(fp,"        %d,\n",cfg->full_length);
        fprintf(fp,"        %d,\n",cfg->row_blocks);
        fprintf(fp,"        %d,\n",cfg->col_blocks);
        fprintf(fp,"        %d,\n",cfg->block_size);
        fprintf(fp,"        %d,\n",block_words);
        fprintf(fp,"        %d,\n",parity);
        fprintf(fp,"        qc_circulant_first_rows_cfg_%d\n",
                cfg->selection);               // maps to first_rows
        fprintf(fp,"    },\n");
    }

    fprintf(fp,"};\n\n");

    fprintf(fp,
    "static const int k_qc_encoder_config_count = "
    "(int)(sizeof(k_qc_encoder_configs) / "
    "sizeof(k_qc_encoder_configs[0]));\n\n");

    fprintf(fp,"#endif\n");

    fclose(fp);
    return 0;
}

/* ============================================================
 * MAIN
 * ============================================================ */

int main(void)
{
    uint64_t *flat_data[MAX_CONFIGS]={0};
    int flat_sizes[MAX_CONFIGS]={0};
    int block_words_array[MAX_CONFIGS]={0};

    for(int c=0;c<CONFIG_COUNT;c++)
    {
        const qc_config_input *cfg=&CONFIGS[c];

        int block_words=0;
        uint64_t *block_rows=
            load_octave_sparse_text_mat(cfg,&block_words);

        if(!block_rows)
            return EXIT_FAILURE;

        block_words_array[c]=block_words;

        int total_words =
            cfg->row_blocks *
            cfg->col_blocks *
            block_words;

        uint64_t *flat=
            malloc(total_words*sizeof(uint64_t));

        if(!flat)
            return EXIT_FAILURE;

        int out_index=0;

        for(int rb=0;rb<cfg->row_blocks;rb++)
        {
            int row_start = rb * cfg->block_size;

            for(int cb=0;cb<cfg->col_blocks;cb++)
            {
                uint64_t *first_row =
                    &block_rows[
                        row_start *
                        cfg->col_blocks *
                        block_words +
                        cb * block_words];

                for(int w=0;w<block_words;w++)
                    flat[out_index++]=first_row[w];
            }
        }

        flat_data[c]=flat;
        flat_sizes[c]=total_words;

        free(block_rows);

        printf("Validated config %d: %s\n",
               cfg->selection,cfg->name);
    }

    if(emit_header(CONFIGS,
                   flat_data,
                   flat_sizes,
                   block_words_array,
                   CONFIG_COUNT)!=0)
        return EXIT_FAILURE;

    printf("Wrote qc_encoder_constants.h\n");

    for(int i=0;i<CONFIG_COUNT;i++)
        free(flat_data[i]);

    return EXIT_SUCCESS;
}